vault/dendron.topic.refactoring.commands.archive-hierarchy.md

```diff
@@ -0,0 +1,36 @@
+---
+id: o350dnypumec34m7xha2g6f
+title: "Dendron: Archive Hierarchy"
+desc: "Move current note and all children under the `archive` hierarchy"
+updated: 1655184589843
+created: 1655184272177
+commandId: dendron.archiveHierarchy
+keybindings:
+  mac: none
+  windows: none
+  linux: none
+---
+
+## Summary
```

Comment:
i don't think the command template was part of this review so going to address this here. currently, the tempmlate is too much work to fill out. it is also duplicate of what is in `package.json`

for templates that require manual work, we should either programatically apply the frontmatter or make it easy to manually update. 

for the `Summary` section, this is usually exerpted in other places so we should keep the information here to a minimum. for the parameters, can we add it under `Details` instead?

- would not include command title, this is already explained by the h1 title
- would not include command id, not relevant to the user
- for keybindings, most commands do not have keybindings. this should be empty if keybindings don't exist

Reply:
I realized our template is part of `org-private`. Here is the [template](https://github.com/dendronhq/org-private/blob/main/notes/templates.command.md?plain=1)

Thoughts on your suggestion:

> the tempmlate is too much work to fill out

On a strictly conceptual level, writing documents is the act of preemptively doing the work for future readers (including the writer themselves) to organize information in a form that is easier to consume. So that the next person down the line doesn't have to put extra effort in gaining a piece of information. With a fixed location of code that holds everything that I included in the template's frontmatter, I consider filling in the information a trivial task even if it has to be done manually.

> it is also duplicate of what is in package.json

That is my exact reasoning for adding them in frontmatter. We can't expect users to read the extension manifest to get information. shouldn't the note mirror what `package.json` has?

Some immediate use cases:
- linking information in Discord
  - linking to `package.json` in source code vs a reference to the section in note
- user not having to wonder where the default keybindings are found in VSCode , for the purpose of customizing them in their `keybindings.json`
  - VSCode doesn't save default keybindings as a file, but rather generates a json on runtime when you run `Preferences: Open Default Keyboard Shortcuts (JSON)`, and you have to search for it there too
- writing doc for commands
  - we were already documenting it in the body of the note. I just formalized it and moved it to the frontmatter and left a fm variable substitution in the body instead.
  - In the future, I plan on adding tooling to programmatically add this. having the frontmatter mirror what's in `package.json` is the natural course of action. (how would we programmatically add it if the information is in a non-fixed location in the body?)

> for templates that require manual work, we should either programatically apply the frontmatter or make it easy to manually update.

That is also my idea, and in the future I'm planning on adding tooling to do so.

> for the Summary section, this is usually exerpted in other places so we should keep the information here to a minimum.

My personal opinion is that the excerpted content should contain the parameters, but I can understand the argument against it.

How about keeping only the one-liner description in `Summary`, and having another section (other than `Details`, as details are more for going into how things work in the command) that only has the parameters.

This way, we can transclude the extra information in places where it makes sense, but also have to freedom to just have a one-liner description transcluded.

> would not include command title, this is already explained by the h1 title

Agreed.

> would not include command id, not relevant to the user

It is absolutely relevant for users.
1. preview now supports command uris (though we haven't advertised this capability publicly yet)
2. you need to know this to add items in `keybindings.json`.

Because of point 2, I was planning to suggest we go even further, and add all available command inputs too, but that remains to be future work for now. (this is user-facing, as this could be used to customize keybindings)

> for keybindings, most commands do not have keybindings. this should be empty if keybindings don't exist

The explicit signal that it is a default doesn't exist adds value in that it gives the reader assurance that the non-existance of it is documented rather than assumed. (we do the same thing with boolean config values in `dendron.yml`)

I would have made it simple and not require all os keys if frontmatter variable substitution can have conditionals, but that remains to be available only for handlebar templates.

If the concern is that it takes too much effort to fill in the values when mostly this information is sparse, the [template](https://github.com/dendronhq/org-private/blob/main/notes/templates.command.md?plain=1) will pre-fill it with "none"

***

Meta thought on this discussion.

I am sensing that we are approaching command documents from a different angle.

I am of the opinion that a documentation for a concrete command be complete in a sense that it contains every single piece of information that is needed, in a reference style, in order to use the command in different ways.

I feel like your approach is to write them in a way that they read naturally in the `topic` context.

Continuing with my proposal in the [other comment](https://github.com/dendronhq/dendron-site/pull/540#discussion_r897494444), I think my approach does have benefits in that it is complete as a reference for a command. If we further divide sections so that we isolate content that we would want to transclude in topics, I think it would make sense for your approach as well to document commands this way.

I can also envision us writing everything that the command document contains where the source code is, and generating API-doc-like documentation for commands automatically in the future, but that remains to be a future work (similar to what we discussed about generating swagger docs for engine API)

Reply:
> On a strictly conceptual level, writing documents is the act of preemptively doing the work for future readers (including the writer themselves) to organize information in a form that is easier to consume. So that the next person down the line doesn't have to put extra effort in gaining a piece of information. With a fixed location of code that holds everything that I included in the template's frontmatter, I consider filling in the information a trivial task even if it has to be done manually.

preemptive work is fine but too much is not. 80/20, keep it simple, etc

some assumptions going into this:
- some standard is better than no standard
- a simple standard that most people will follow is better than a complicated standard that doesn't have compliance
- we should design for use cases that have come up vs use cases we think will come up (help with the previous two points)

the biggest issue with standards are that they are not followed. as we create a standard, our number one priority is creating something that will be easy for people to follow and maintain over time. the problem with specifying all this information is that it will all have to be maintained and can quiickly become overwhelming

> Some immediate use cases: linking information in Discord, user not having to wonder where the default keybindings are found in VSCode 

- these aren't exactly common use cases on discord (eg. what is dendronId). and the command name is usually good enough to get people going

> we were already documenting it in the body of the note. I just formalized it and moved it to the frontmatter and left a fm variable substitution in the body instead.

We're not though.  An example of a current command: [[Publish Pod|dendron://dendron.dendron-site/dendron.ref.commands#publish-pod]]
  ```md
  ### Publish Pod

  - shortcuts: none

  Publish your note to a different format/location
  ```

This is on the light site in terms of docs. This is why I'm suggesting something with more information, like [[Apply Template|dendron://dendron.dendron-site/dendron.topic.templates.commands.apply-template]]

> This way, we can transclude the extra information in places where it makes sense, but also have to freedom to just have a one-liner description transcluded.

this is good for me. in terms of tooling, we should also be able to transclude frontmatter properties so that we can just transclude the `desc` field in the future

> It is relevant for users. (in regards to command id)

for now, i would point people to `package.json` - this currently adds more maintenance. we already struggle with keeping docs up to date, anything that requires continuous maintenance is going to be hard to support

> I would have made it simple and not require all os keys if frontmatter variable substitution can have conditionals, but that remains to be available only for handlebar templates.

handlebars is now the default going forward :)
for keyboard shortcuts, it should either be `none` or it should show keyboard shortcuts that are platform specific

> I am of the opinion that a documentation for a concrete command be complete in a sense that it contains every single piece of information that is needed, in a reference style, in order to use the command in different ways.

yep, agreed. see my previous comment about trying out the horizontal approach with commands but using the vertical for everything else

Reply:
> the biggest issue with standards are that they are not followed. as we create a standard, our number one priority is creating something that will be easy for people to follow and maintain over time. the problem with specifying all this information is that it will all have to be maintained and can quiickly become overwhelming

I don't think we are in disagreement with that. Rather, we are in disagreement that using my proposed template is too much work. How about we let other folks who regularly write documents chime in?

We have many internal standard documents that have much more frontmatter in them that, over time, we have accepted to be a fair and manageable amount of work to fill in, that we have been creating and maintaining in much higher frequency and volume than we create and edit documents about commands. I am still not convinced that the amount of work I am proposing is too much compared to a lot of what we are doing with internal documents.

My gravitation towards a more structured frontmatter for these information as a writer of document is developer experience; the affordance of the frontmatter that comes with them being explicit. An empty plastic cup on a ledge of a staircase give people an affordance that it is a place to throw their cups away. An explicit frontmatter in the template gives doc writers the affordance that you are expected to document those information. It is acting as a signifier.

Assuming that we are going with my template, 
1. how frequent are we going to change the value of the frontmatters (`dendronId` and default keybingings) once a new note for a command was created and the values were first inputted?
    - Realistically, the values in the frontmatter will be set at the point the command is created and documented, and very rarely change since they will be breaking changes in the user experience. If they do indeed change, I doubt that the fact that you have to update the frontmatter will slip since introducing a breaking change is a significant enough event that it calls for an official deprecation notice in our release note and we should make sure that everything is in place before we do the announcement to begin with.
2. Is it in the territory where we have to worry about this not being a scalable solution for documenting commands?
3. Comparing it to the relaxed lighter solution of just keeping them as a short mention directly in the body, how much extra work are we actually doing? 
    - Going back to your argument of this being too much work for it to be maintained up to date, I would actually say putting it in the frontmatter makes it much more maintainable since it's immediately visible once the document is opened, and it signals that somewhere in the body we are using that information. 
    - Keeping the frontmatter up to date versus keeping the direct mentions of them in the body is roughly the same if we assume that in both cases the maintainer is _aware that we need to maintain that up to date_ (which is not a given as-is, but I'm claiming that the frontmatter improves the chance)

> these aren't exactly common use cases on discord (eg. what is dendronId). and the command name is usually good enough to get people going

I think there is a misunderstanding. I am not saying the command id is a command way to describe a command in support cases. Nor am I saying the command ID is a useful keyword for looking up the command. I am saying that this is a _valid piece of information that is exposed to the user directly and used commonly, and would benefit from being in the document so that we can refer to them with ease_. I have had many instances of responding with, and have seen users wonder what the exact command ids are. So far I have resorted to just answering them after customizing my keybindings for their use case and copy pasting what I've come up with.

> We're not though. An example of a current command: [[Publish Pod|dendron://dendron.dendron-site/dendron.ref.commands#publish-pod]]

I am referring to the part of the doc you linked that says `- shortcuts: none`. I simply made them into a frontmatter property and made them more explicit by iterating on all platforms. And they are substituted back in the same place. so structurally this change is transparent to the reader.

> for now, i would point people to package.json - this currently adds more maintenance. we already struggle with keeping docs up to date, anything that requires continuous maintenance is going to be hard to support

Again, I am claiming that putting this information in frontmatter is an improvement over our as-is state. Even after stripping away my claim that having it in frontmatter makes it more visible for maintainers, maintaining what we currently have versus maintaining the same thing in frontmatter doesn't impact the difficulty of continuously maintaining them up to date. It is certain that having them in the frontmatter doesn't make it worse. It's just a different thing we have to maintain, and that concern is orthogonal to the issue of where we are documenting them.

> handlebars is now the default going forward :)

Are you saying we can use handlebars construct in frontmatter variable substitution today? I don't think that is the case yet. If you check my template again, the frontmatter I proposed to add to command notes are going to a section of the body, in a fixed location, without any steps needed for the maintainer to modify the section because the template contains an escaped frontmatter variable substitution expression under `## Summary` (which I will modify to be a separate section to isolate it from the summary).

> for keyboard shortcuts, it should either be none or it should show keyboard shortcuts that are platform specific

Frontmatter variable substitution is not straightforward today if the type signature of the frontmatter we are referring to is a union because like I mentioned, we can't use conditionals for frontmatter variable substitution. That is one reason I am explicitly iterating on all three platforms in the frontmatter. Another reason would be the one that I keep bringing up, that being explicit adds value over assuming non-existence of a certain property.

(of course, this problem only is relevant in the assumption that we are going to keep keybindings in the frontmatter.)

Reply:
removed all custom frontmatter and simplified [templates.command](https://github.com/dendronhq/org-private/blob/main/notes/templates.command.md?plain=1)